### PR TITLE
Unlock collateral and borrow debt

### LINKED ISSUE #3: Implement unlock ATOM message handler

### LINKED ISSUE #5: Implement borrow MEI message handler

### LINKED ISSUE #7: Implement calculate collateral ratio function

### PR DESCRIPTION
**Pull request by tharamalai**

### CODE REVIEW

File: x/meicdp/handler.go [hunk 1]
```diff
@@ -19,23 +23,122 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
 
 		case MsgLockCollateral:
 			return handleMsgLockCollateral(ctx, keeper, msg)
 
+		case MsgUnlockCollateral:
+			msgCount := keeper.GetMsgCount(ctx)
+
+			// Setup oracle request
+			bandChainID := BandChainID
+			port := MeiCdpPort
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: Can't we use `ModuleName` here?

File: x/meicdp/handler.go [hunk 2]
```diff
@@ -19,23 +23,122 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+			return nil, sdkerrors.Wrapf(
+				sdkerrors.ErrUnknownRequest,
+				"cannot unmarshal oracle packet data",
+			)
 
 		case MsgLockCollateral:
 			return handleMsgLockCollateral(ctx, keeper, msg)
 
+		case MsgUnlockCollateral:
+			msgCount := keeper.GetMsgCount(ctx)
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: Let's call this `msgID`

File: x/meicdp/handler.go [hunk 3]
```diff
@@ -19,23 +23,122 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+					types.ErrRequestOracleData,
+					"error while request oracle data: %v",
+					err,
+				)
+			}
+
+			return &sdk.Result{Events: ctx.EventManager().Events().ToABCIEvents()}, nil
+
+		case MsgBorrowDebt:
+			msgCount := keeper.GetMsgCount(ctx)
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: This is duplicate of the above. Can't we handle them together?

File: x/meicdp/handler.go [hunk 4]
```diff
@@ -19,23 +23,122 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+					err,
+				)
+			}
+
+			return &sdk.Result{Events: ctx.EventManager().Events().ToABCIEvents()}, nil
+
 		case MsgReturnDebt:
 			return handleMsgReturnDebt(ctx, keeper, msg)
 
 		case MsgSetSourceChannel:
```
Comment (defi-pizza) [inline x/meicdp/handler.go:46]: Add TODO. Check permission

File: x/meicdp/handler.go [hunk 5]
```diff
@@ -19,23 +23,122 @@ func NewHandler(keeper Keeper) sdk.Handler {
 		case channeltypes.MsgPacket:
 			var responseData oracle.OracleResponsePacketData
 			if err := types.ModuleCdc.UnmarshalJSON(msg.GetData(), &responseData); err == nil {
-				fmt.Println("I GOT DATA", responseData.Result, responseData.ResolveTime)
-				// handleOraclePacket(ctx, keeper, responseData)
+				err := handleOracleRespondPacketData(ctx, keeper, responseData)
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: Will this fail the whole transaction if an error occurs?

File: x/meicdp/handler.go [hunk 6]
```diff
@@ -117,3 +239,215 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+			dataReq.Port,
+		)
+	}
+
+	packet := oracle.NewOracleRequestPacketData(
+		dataReq.ClientID, dataReq.OracleScriptID, dataReq.Calldata,
+		dataReq.AskCount, dataReq.MinCount,
+	)
+
+	err = keeper.ChannelKeeper.SendPacket(ctx, channel.NewPacket(packet.GetBytes(),
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: can't we return here?

**Review by defi-pizza — CHANGES_REQUESTED** [review]

File: x/meicdp/handler.go (hunk 1 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 2 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 4 above)
Comment (tharamalai) [reply x/meicdp/handler.go:46]: added

File: x/meicdp/handler.go (hunk 6 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 3 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: Added new handler function to handle together

File: x/meicdp/handler.go [hunk 7]
```diff
@@ -117,3 +180,259 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+		err := handleMsgUnlockCollateral(ctx, keeper, msg, collateralPrice)
+		if err != nil {
+			return err
+		}
+
+	case types.MsgBorrowDebt:
+		err := handleMsgBorrowDebt(ctx, keeper, msg, collateralPrice)
+		if err != nil {
+			return err
+		}
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: default should error.

File: x/meicdp/handler.go [hunk 8]
```diff
@@ -117,3 +180,259 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+
+// handleMsgBorrowDebt handles the borrow debt message after receives oracle packet
+func handleMsgBorrowDebt(ctx sdk.Context, keeper Keeper, msg types.MsgBorrowDebt, collateralPrice uint64) error {
+	cdp := keeper.GetCDP(ctx, msg.Sender)
+
+	borrowAmount := sdk.NewCoin(types.MeiUnit, sdk.NewInt(int64(msg.Amount)))
+	borrowAmountCoins := sdk.NewCoins(borrowAmount)
+
+	// Accumurate debt on CDP
+	borrowAmountInt := new(big.Int).SetUint64(msg.Amount)
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: We should convert this to `sdk.Coin` or `sdk.Coins` and use comparison function from there. Repeat this everywhere.

**Review by defi-pizza — CHANGES_REQUESTED** [review]

File: x/meicdp/handler.go (hunk 7 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 5 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: yes, this err will fail the whole transaction and didn't block other action

File: x/meicdp/handler.go (hunk 8 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: Converted to coin type

File: x/meicdp/handler.go [hunk 9]
```diff
@@ -117,3 +195,263 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+			types.ErrRequestOracleData,
+			"error while request oracle data: %v",
+			err,
+		)
+	}
+
+	return nil
+}
+
+func handleOracleRespondPacketData(ctx sdk.Context, keeper Keeper, packet oracle.OracleResponsePacketData) error {
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: Response

File: x/meicdp/handler.go [hunk 10]
```diff
@@ -117,3 +195,263 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+	if err != nil {
+		return sdkerrors.Wrapf(
+			types.ErrMsgNotFound,
+			"cannot get stored message",
+		)
+	}
+
+	switch msg := msg.(type) {
+	case MsgUnlockCollateral:
+		err := handleMsgUnlockCollateral(ctx, keeper, msg, collateralPrice)
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: return

File: x/meicdp/handler.go [hunk 11]
```diff
@@ -117,3 +195,263 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+		if err != nil {
+			return err
+		}
+
+	case MsgBorrowDebt:
+		err := handleMsgBorrowDebt(ctx, keeper, msg, collateralPrice)
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: return

File: x/meicdp/handler.go [hunk 12]
```diff
@@ -117,3 +195,263 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+		}
+
+	default:
+		return sdkerrors.Wrapf(
+			types.ErrInvalidMsgType,
+			fmt.Sprintf("invalid message type: %T", msg),
+		)
+	}
+
+	return nil
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: this is not needed

File: x/meicdp/handler.go [hunk 13]
```diff
@@ -117,3 +195,263 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
...
+	minimumDebtAmount := new(big.Int).SetUint64(0)
+	if debtAmount.Cmp(minimumDebtAmount) > 0 {
+
+		collateralRatioFloat := calculateCollateralRatioOfCDP(cdp, collateralPrice, AtomMultiplier)
+		minimunRatio := new(big.Float).SetFloat64(MinimumCollateralRatio)
+		collateralRatio, _ := collateralRatioFloat.Float64()
+		if collateralRatioFloat.Cmp(minimunRatio) == -1 {
+			return sdkerrors.Wrapf(
+				types.ErrTooLowCollateralRatio,
+				fmt.Sprintf("collateral rate is too low. (%f%)", collateralRatio),
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: not work

**Review by defi-pizza — CHANGES_REQUESTED** [review]

File: x/meicdp/handler.go (hunk 9 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 10 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 12 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/handler.go (hunk 13 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

**Pull request merged**